sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -165,6 +165,85 @@ namespace sycl::ext::oneapi::experimental::matrix {
+}
+```
+IMPORTANT: In the current implementation, only the subgroup scope is supported.  
+
+#### Element Indexing and Piece-Wise Operations 
+Besides matrix multiply and add, matrices are used in linear and non linear piece-wise operations. Activation functions are an example of element-wise operations. They can be linear like `ReLU` that, for each value `z`, returns the maximum between `z` and zero, or non linear like `Sigmoid` that calculates `1/(1+ exp(-z))`. Quantization that is needed for conversion between low precision types like `int8_t` and `fp32` uses piece-wise operations. For instance, quantized GEMM for `int8_t` is calculated using `A*B + sum_rows_A + sum_cols_B + scalar_zero_point`. `sum_rows_A` and `sum_cols_B` do not operate on elements of the matrix but on pieces: row in `sum_rows_A` and columns in `sum_cols_B`.
+
+We explored multiple options to enable this feature in the matrix interface: 1) Allowing non-restrictive element indexing on the matrix elements would result into slow indexing on the GPU, 2) Operator overloading can represent only element-wise operations and not the operations on pieces (row, column, diagonal, etc) of the matrix. 3) Providing specific functions for these piece-wise operations can resolve some of the functions we know of today like the ones involved in quantization but it is not general to any problem that may occur in the future. 
+
+In order to be able to perform any piece-wise operation in a general and in an efficient way, we provide a mapping conversion function from the matrix domain that is owned by a group of work items to the portion that is owned by a work item. Besides, the WI data to joint matrix mapping coordinates information must be known. 
+
+Nvidia wmma interface added a new member to `fragment` class to designate the WI owned part of the matrix. 
+While this provides fast element indexing on the GPU compared to the non-restrictive option, the user does not know the mapping of the owned data to the original matrix. 
+ However using the `mma` ptx instructions as opposed to the `wmma` ptx instructions the mapping is known. Knowing this mapping is important for the user to implement new operations like sum of rows of a matrix for quantized algorithms.
+
+##### Solution: Explicit conversion with mapping from SIMD to SPMD
+We introduce a new function `get_wi_data` that provides a view of the portion of the matrix that is owned by the current WI.
+
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, size_t NumRows, size_t NumCols,
+          matrix_layout Layout = matrix_layout::row_major,
+          typename Group = sycl::sub_group>
+struct joint_matrix {
+   joint_matrix(Group sg) {…}   
+   wi_data<T, NumRows, NumCols, Layout, Group> get_wi_data() {
+     return wi_data<T, NumRows, NumCols, Layout, Group>(*this);
+   }
+};
+template <typename T, size_t NumRows, size_t NumCols, matrix_layout Layout, typename Group>
```

Comment:
Do `wi_data` and `wi_element` really need all these template parameters?  It seems like it would be easier to use if the only template parameter was `T`.  It seems like the other template parameters are only there because there is a private data member `M` (a reference to matrix).  However, you only seem to use `M.spvm` in the function bodies.  Could you instead just store the `spvm` member directly in `wi_data` and `wi_element`?

Reply:
Can you elaborate on your suggestion? I get what you want to do but did not get the how.

Reply:
Sorry for the delayed response.  I was OOO for about 2 months and just got back recently.

I was thinking that these types could be simplified to have fewer template parameters, so the goal would be to have an API like this:

```
template <typename T>
class wi_data {
  size_t length();
  wi_element<T> operator[](size_t i);
};

template <typename T>
class wi_element {
  operator T();
  wi_element &operator=(const T &rhs);
};
```

Very roughly, I was thinking that you could accomplish this by changing the private data member included in `wi_data` and `wi_element`.  Currently, these both contain a reference to the joint matrix `M`.  However, it seems like they only need to use `M.spvm`.  Therefore, could you change the implementation to hold just the `spvm` like:

```
template <typename T>
class wi_data {
  /* not sure what type */ spvm;
public:
  size_t length() {return __spirv_JointMatrixWorkItemLengthINTEL(spvm);}
  wi_element<T> operator[](size_t i) {
    return wi_element<T>(spvm, i);
  }
};

template <typename T>
class wi_element {
  /* not sure what type */ spvm;
  std::size_t idx;

public:
  operator T() {
    return __spirv_VectorExtractDynamic(spvm, idx);
  }
  wi_element &operator=(const T &rhs) {
    M.spvm = __spirv_VectorInsertDynamic(spvm, rhs, idx);
    return *this;
  }
};
```

Reply:
@gmlueck 
It looks like spvm type also needs these template parameters. SO I don't think we can reduce them:
 __spv::__spirv_JointMatrixINTEL<
      T, NumRows, NumCols, spv_matrix_layout_traits<Layout>::value> *spvm;

Reply:
That's too bad.